contracts/extensions/contracts/RecipientsExtension.sol

```diff
@@ -0,0 +1,329 @@
+                emit Registered(recipientId, extendedData);
+
+                recipientsCounter++;
+            } else {
+                uint8 currentStatus = _getUintRecipientStatus(recipientId);
+                if (currentStatus == uint8(Status.Accepted) || currentStatus == uint8(Status.InReview)) {
+                    // recipient updating accepted application
+                    _setRecipientStatus(recipientId, uint8(Status.Pending));
+                } else if (currentStatus == uint8(Status.Rejected)) {
+                    // recipient updating rejected application
+                    _setRecipientStatus(recipientId, uint8(Status.Appealed));
+                }
+                emit UpdatedRegistration(recipientId, data, _sender, _getUintRecipientStatus(recipientId));
+            }
+
+            _recipientIds[i] = recipientId;
+        }
+    }
+
+    /// @notice Extract recipient and metadata from the data.
+    /// @param _data The data to be decoded
+    /// @param _sender The sender of the transaction
+    /// @return recipientId The ID of the recipient
+    /// @return isUsingRegistryAnchor A flag to indicate whether to use the registry anchor or not
+    /// @return metadata The metadata of the recipient
+    function _extractRecipientAndMetadata(bytes memory _data, address _sender)
+        internal
+        view
+        virtual
+        returns (address recipientId, bool isUsingRegistryAnchor, Metadata memory metadata)
+    {
+        (address _recipientIdOrRegistryAnchor, Metadata memory _metadata) = abi.decode(_data, (address, Metadata));
+
+        metadata = _metadata;
+
+        // If the registry anchor is not the zero address check authorization
+        // Anchor can never be zero, so if zero, set '_sender' as the recipientId
+        if (_recipientIdOrRegistryAnchor != address(0)) {
+            if (!_isProfileMember(_recipientIdOrRegistryAnchor, _sender)) {
+                revert UNAUTHORIZED();
+            }
+
+            isUsingRegistryAnchor = true;
+            recipientId = _recipientIdOrRegistryAnchor;
+        } else {
+            // Using 'isUsingRegistryAnchor' default value (false)
+            recipientId = _sender;
+        }
+    }
+
+    /// @notice Get the recipient details.
+    /// @param _recipientId Id of the recipient
+    /// @return Recipient details
+    function _getRecipient(address _recipientId) internal view virtual returns (Recipient memory) {
+        return _recipients[_recipientId];
+    }
+
+    /// @notice Set the recipient status.
+    /// @param _recipientId ID of the recipient
+    /// @param _status Status of the recipient
+    function _setRecipientStatus(address _recipientId, uint256 _status) internal virtual {
+        // Get the row index, column index and current row
+        (uint256 rowIndex, uint256 colIndex, uint256 currentRow) = _getStatusRowColumn(_recipientId);
+
+        // Calculate the 'newRow'
+        uint256 newRow = currentRow & ~(15 << colIndex);
```

Comment:
Just a personal preference: for this kind of bit operation I think using hex, 0xff instead of 15, is more readable, but both are ok.